src/stores/widgets/WidgetLayoutStore.ts

```diff
@@ -0,0 +1,502 @@
+    // to 0-100 and may have minimums imposed on it.
+    height?: number;
+
+    // TODO: [Deferred] Maximizing (fullscreen) widgets by default.
+}
+
+interface IWidgetLayouts {
+    [widgetId: string]: IStoredLayout;
+}
+
+interface ILayoutStateEvent {
+    // TODO: [Deferred] Forced layout (fixed with no changes)
+
+    // The widget layouts.
+    widgets: IWidgetLayouts;
+}
+
+interface ILayoutSettings extends ILayoutStateEvent {
+    overrides?: string; // event ID for layout state event, if present
+}
+
+// Dev note: "Pinned" widgets are ones in the top container.
+export const MAX_PINNED = 3;
+
+// These two are whole percentages and don't really mean anything. Later values will decide
+// minimum, but these help determine proportions during our calculations here. In fact, these
+// values should be *smaller* than the actual minimums imposed by later components.
+const MIN_WIDGET_WIDTH_PCT = 10; // 10%
+const MIN_WIDGET_HEIGHT_PCT = 2; // 2%
+
+export class WidgetLayoutStore extends ReadyWatchingStore {
+    private static internalInstance: WidgetLayoutStore;
+
+    private byRoom: {
+        [roomId: string]: {
+            // @ts-ignore - TS wants a string key, but we know better
+            [container: Container]: {
+                ordered: IApp[];
+                height?: number;
+                distributions?: number[];
+            };
+        };
+    } = {};
+
+    private pinnedRef: string;
+    private layoutRef: string;
+
+    private constructor() {
+        super(defaultDispatcher);
+    }
+
+    public static get instance(): WidgetLayoutStore {
+        if (!WidgetLayoutStore.internalInstance) {
+            WidgetLayoutStore.internalInstance = new WidgetLayoutStore();
+        }
+        return WidgetLayoutStore.internalInstance;
+    }
+
+    public static emissionForRoom(room: Room): string {
+        return `update_${room.roomId}`;
+    }
+
+    private emitFor(room: Room) {
+        this.emit(WidgetLayoutStore.emissionForRoom(room));
+    }
+
+    protected async onReady(): Promise<any> {
+        this.updateAllRooms();
+
+        this.matrixClient.on("RoomState.events", this.updateRoomFromState);
+        this.pinnedRef = SettingsStore.watchSetting("Widgets.pinned", null, this.updateFromSettings);
+        this.layoutRef = SettingsStore.watchSetting("Widgets.layout", null, this.updateFromSettings);
+        WidgetStore.instance.on(UPDATE_EVENT, this.updateFromWidgetStore);
+    }
+
+    protected async onNotReady(): Promise<any> {
+        this.byRoom = {};
+
+        SettingsStore.unwatchSetting(this.pinnedRef);
+        SettingsStore.unwatchSetting(this.layoutRef);
+        WidgetStore.instance.off(UPDATE_EVENT, this.updateFromWidgetStore);
+    }
+
+    private updateAllRooms = () => {
+        this.byRoom = {};
+        for (const room of this.matrixClient.getVisibleRooms()) {
+            this.recalculateRoom(room);
+        }
+    };
+
+    private updateFromWidgetStore = (roomId?:string) => {
```

Comment:
```suggestion
    private updateFromWidgetStore = (roomId?: string) => {
```

There must be linting rule for this...

Reply:
https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/type-annotation-spacing.md

Reply:
Normally it catches this...

Reply:
Ah, I guess it's not enabled, I'll add it to our config.

Reply:
https://github.com/matrix-org/eslint-config-matrix-org/pull/4